modules/sr/robot3/arduino_devices.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import annotations
+
+import abc
+import logging
+from typing import Optional
+
+from controller import (
+    LED,
+    Robot,
+    TouchSensor,
+    DistanceSensor as WebotsDistanceSensor,
+)
+from sr.robot3.utils import map_to_range, get_robot_device
+from controller.device import Device
+from sr.robot3.randomizer import add_jitter
+from sr.robot3.output_frequency_limiter import OutputFrequencyLimiter
+
+LOGGER = logging.getLogger(__name__)
+
+
+class PinDevice(abc.ABC):
```

Comment:
If we're going to use this approach (devices separate from pins), which I admit is probably a better simulation of the real robot, then I think we should model the simulation as such. That is: allow for any device being used on any pin and have them communicate in a way that reflects reality.

Specifically I think we should have _only_ analogue-read and digital-write operations at the device level. We can do the analogue-to-digital conversion consistently in the `Pin` type, with a suitable threshold.

This approach avoids all the quandaries of whether a given type should implement a given interface or whether returning analogue zero from a microswitch is correct. If we're modelling these as separate devices then modelling them as _real_ analogue devices is the way to go.

Reply:
The purpose of the simulation is to model the real world in a simplified way such that the behaviour of a system can be understood. The simplification here is to make is clear how a device should be used; a distance sensor is analog, a micro-switch is digital.

A warning is now printed when one of the default methods are used to warn competitors that this is not how the device is meant to be used.